### ISSUE TITLE
"Ignoring option use-resolvconf" on Fedora 35

### ISSUE THREAD

**Issue by reetis** [adrienverge/openfortivpn#957]
The message `Ignoring option "use-resolvconf".` is shown on Fedora 35 when using `use-resolvconf = 1` in config file. It is working on Fedora 34. Package version `openfortivpn-1.17.0-2.fc35`.

**Comment by DimitriPapadopoulos** [comment]
Ideally, you should open a ticket here
https://bugzilla.redhat.com/buglist.cgi?component=openfortivpn&list_id=12270035&product=Fedora

This does not look like an openfortivpn issue, but a packaging issue.

**Comment by DimitriPapadopoulos** [comment]
I'll try to have a look nevertheless.

**Comment by reetis** [comment]
Opened a bug report in Red Hat Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=2024565

**Comment by DimitriPapadopoulos** [comment]
The error message is emitted from:
https://github.com/adrienverge/openfortivpn/blob/d1b2801b6e866411398414be199c7648eb3e620b/src/config.c#L363-L364

https://github.com/adrienverge/openfortivpn/blob/d1b2801b6e866411398414be199c7648eb3e620b/src/config.c#L373-L375

It is emitted because **`HAVE_RESOLVCONF`** is not defined.

**Comment by DimitriPapadopoulos** [comment]
Now **`HAVE_RESOLVCONF`** is set at build time by autoconf, so it's a build time, hence a packaging issue.

It looks like _openfortivpn_ has been configured with `--with-resolvconf=DISABLE`.

**Comment by DimitriPapadopoulos** [comment]
No, all packages have been built **without** `--with-resolvconf=DISABLE` starting with Fedora 33:
* branch `f32` has been built **with** `--with-resolvconf=DISABLED --disable-resolvconf`:
  https://src.fedoraproject.org/rpms/openfortivpn/blob/f32/f/openfortivpn.spec#_28
* branch `f33` has been built **without** `--with-resolvconf=DISABLED --disable-resolvconf`:
  https://src.fedoraproject.org/rpms/openfortivpn/blob/f33/f/openfortivpn.spec#_28
* branch `f34` has been built **without** `--with-resolvconf=DISABLED --disable-resolvconf`:
  https://src.fedoraproject.org/rpms/openfortivpn/blob/f34/f/openfortivpn.spec#_29
* branch `f35` has been built **without** `--with-resolvconf=DISABLED --disable-resolvconf`:
  https://src.fedoraproject.org/rpms/openfortivpn/blob/f35/f/openfortivpn.spec#_29
* branch `rawhide` has been been built **without** `--with-resolvconf=DISABLED --disable-resolvconf`:
  https://src.fedoraproject.org/rpms/openfortivpn/blob/rawhide/f/openfortivpn.spec#_39

**Comment by DimitriPapadopoulos** [comment]
I had a look at the build logs of `f34` and `f35`:
https://kojipkgs.fedoraproject.org//packages/openfortivpn/1.17.0/2.fc34/data/logs/x86_64/build.log
https://kojipkgs.fedoraproject.org//packages/openfortivpn/1.17.0/2.fc35/data/logs/x86_64/build.log

They are different indeed:
* for `f34`:
  ```
  checking for resolvconf... /usr/sbin/resolvconf
  configure: RESOLVCONF_PATH... /usr/sbin/resolvconf
  configure: HAVE_RESOLVCONF... 1
  checking for /usr/sbin/resolvconf... yes
  configure: USE_RESOLVCONF... 0
  ```
* for `f35`:
  ```
  checking for resolvconf... DISABLED
  configure: RESOLVCONF_PATH... DISABLED
  configure: HAVE_RESOLVCONF... 0
  configure: USE_RESOLVCONF... 0
  ```

We haver to understand why `configure` results in a different configurations.

**Comment by DimitriPapadopoulos** [comment]
The reason has to be here:
https://github.com/adrienverge/openfortivpn/blob/d1b2801b6e866411398414be199c7648eb3e620b/configure.ac#L345-L346

I believe `RESOLVCONF_PATH` is set to `DISABLED` because `resolvconf` is not available at build time - it was available for `f34` but not `f35`.

**Comment by DimitriPapadopoulos** [comment]
For some reason, `resolvconf` was available by default in `f34` but not in `f35`. The packager needs to force `resolvconf` as a dependency at build time and run time, by changing:
```spec
BuildRequires: make
BuildRequires:  gcc autoconf automake
BuildRequires:  pkgconfig(openssl)
BuildRequires:  pkgconfig(systemd)
Requires:       ppp
```
into:
```spec
BuildRequires:  make
BuildRequires:  gcc autoconf automake
BuildRequires:  pkgconfig(openssl)
BuildRequires:  pkgconfig(systemd)
BuildRequires:  resolvconf
Requires:       ppp resolvconf
```

So:
* `resolvconf` support needs to be enabled at build (configure) time, either by installing a package that provides `resolvconf` as a prerequisite, or by hardwiring the path to `resolvconf` with `--with-resolvconf=</path/to/resolvconf>`.
* `resolvconf` needs to be available at run time, by installing a package that provides `resolvconf` as a prerequisite.

**Comment by nubitic-admin** [comment]
> The reason has to be here:
> 
> https://github.com/adrienverge/openfortivpn/blob/d1b2801b6e866411398414be199c7648eb3e620b/configure.ac#L345-L346
> 
> I believe `RESOLVCONF_PATH` is set to `DISABLED` because `resolvconf` is not available at build time - it was available for `f34` but not `f35`.

I happened to realize that on my case it was 'cause the final destination for the /usr/sbin/resolvconf symlink (wich is /usr/sbin/resolvconf.openresolv) had no execute permission. I no longer see the warning <Ignoring option "use-resolvconf"> however I'm still unable to get those DNS working. I'm working on Fedora 36.

**Comment by DimitriPapadopoulos** [comment]
Perhaps openresolv is broken on Fedora 36 - which could explain why `/usr/sbin/resolvconf.openresolv` was not executable. You may try to reinstall (`dnf reinstall`) opensresolv.

**Comment by DimitriPapadopoulos** [comment]
That said, I believe the packages still need to be fixed:
* for `f36`:
  https://kojipkgs.fedoraproject.org//packages/openfortivpn/1.17.0/4.fc36/data/logs/x86_64/build.log
  ```
  checking for resolvconf... DISABLED
  configure: RESOLVCONF_PATH... DISABLED
  configure: HAVE_RESOLVCONF... 0
  configure: USE_RESOLVCONF... 0
  ```

**Comment by DimitriPapadopoulos** [comment]
@nubitic-admin It is unclear which version you are using. Are you building from source or using the RPM package?

**Comment by DimitriPapadopoulos** [comment]
I had a look at the [openresolv package for Fedora 36:](https://packages.fedoraproject.org/pkgs/openresolv/openresolv/fedora-36.html), it looks broken:
```
$ rpm -qlpv ./openresolv-3.12.0-3.fc36.noarch.rpm 
[...]
-rw-r--r--    1 root     root                        0 janv. 20 23:45 /usr/sbin/resolvconf
-rw-r--r--    1 root     root                    23971 janv. 20 23:45 /usr/sbin/resolvconf.openresolv
[...]
$ 
```

**Comment by reetis** [comment]
I might be mistaken, but openresolv just writes DNS entries to `/etc/resolv.conf` which is ignored on Fedora.
On a clean Fedora 36 installation, `/usr/sbin/resolvconf` is provided by systemd-resolved (https://packages.fedoraproject.org/pkgs/systemd/systemd-resolved/fedora-36.html)

**Comment by nubitic-admin** [comment]
> @nubitic-admin It is unclear which version of openfortivpn you are using. Are you building from source or using the RPM package?

Hi, I share with you the process I followed:
1. I tried with yum install openfortivpn
2. At connection attempt I received this warning <Ignoring option "use-resolvconf">
3. I removed it using yum remove openfortivpn
4. I proceeded to download the source code and compile it, I installed openresolv using yum install openresolv
5. At first attempt of compilation I noticed it wasn't detecting openresolv, I managed to find the reason I shared previously
6. After granting execute permissions on /usr/sbin/resolvconf.openresolv compilation was ok, I proceeded to install it
7. At connection attempt I no longer saw the warning but DNSs still had not working

After spending a couple of hours reading on line y tried the following
1. Uninstalled the compiled version
2. Reinstalled using yum install openfortivpn
3. Droped a conf file in /etc/systemd/resolved.conf.d/ indicating DNSs IP for the vpn and the domain name it resolves
4. Connect to the vpn using sudo openfortivpn -c myclient-vpn-config.conf without use-resolvconf option

And voila, DNSs now working.

Here the template of the conf file I used
```
[Resolve]
Cache=yes
DNS=10.10.XX.XX 10.10.YY.YY
Domains=miclient.maindomain
```
All the subdomains bellow miclient.maindomain are now resolved correctly

I glanced what this conf file changes on /etc/resolv.conf and only appends this to the end of the file
```
search miclient.maindomain --
```

I ended up writing a bash script that perform both things, drop the conf file in /etc/systemd/resolved.conf.d/ and connect to the vpn

**Comment by DimitriPapadopoulos** [comment]
The whole point of **resolvconf**, whether the [original Debian implementation](https://salsa.debian.org/debian/resolvconf) or the [openresolv](https://roy.marples.name/projects/openresolv/) one, is to do the right thing. My understanding is that it knows when to modify `/etc/resolv.conf` directly, on older systems, or when to call `systemd-resolve` on recent systemd-based systems. But I may be wrong.

If that's not the case, the only solution would be to use systemd-resolve directly, or perhaps call [vpnc-scripts](https://gitlab.com/openconnect/vpnc-scripts) instead, which is already used by [OpenConnect](https://gitlab.com/openconnect/openconnect) and does the right thing on a variety of Linux distributions and macOS.

**Comment by ddobbelaere** [comment]
> I might be mistaken, but openresolv just writes DNS entries to `/etc/resolv.conf` which is ignored on Fedora. On a clean Fedora 36 installation, `/usr/sbin/resolvconf` is provided by systemd-resolved (https://packages.fedoraproject.org/pkgs/systemd/systemd-resolved/fedora-36.html)

I can confirm that this is the case. Since Fedora 36, the DNS entries added to `/etc/resolv.conf` are ignored (it used to work on Fedora 35 though, for some strange reason that escapes me).

**Comment by DimitriPapadopoulos** [comment]
I know that since Fedora 36, the DNS entries added to `/etc/resolv.conf` are ignored.

What I do not know, is whether `resolvconf` only modifies `/etc/resolv.conf`. I was under the impression it does the "right thing", whatever the DNS backend. If it does not, `resolvconf` is clearly outdated and should not be used any more. On the other hand, a third version of `resolvconf` is (or was) provided by systemd itself, specifically for the systemd backend. That could be a temporary solution. For example on ArchLinux:
https://archlinux.org/packages/?name=systemd-resolvconf
Is the `resolvconf` from systemd available on Fedora?

I still believe the best long-term solution would be to switch to [vpnc-scripts](https://gitlab.com/openconnect/vpnc-scripts), as used by [OpenConnect](https://gitlab.com/openconnect/openconnect).

**Comment by reetis** [comment]
`resolvconf` provided by systemd works correctly with openfortivpn. This is a problem with a Fedora package, since you can't force usage of `resolvconf` without building `openfortivpn` yourself.

**Comment by DimitriPapadopoulos** [comment]
Yes, the Fedora package needs to be fixed with that respect. I suggest you add comments about the Fedora package to this existing open issue:
https://bugzilla.redhat.com/show_bug.cgi?id=2024565

**Comment by aiace9** [comment]
I am having the same problem with fedora 36, but I still have to decide if it makes more sense to change the fedora build or consider that under systemd `resolveconf` is only a symlink to `resolvectl` hence detect that that binary exists.  I think the latter has the advantage of being a more general solution.

**Comment by yurigba** [comment]
I confirm the issue with the fedora package is still on fedora 38 (NOTE: the error in the image is unrelated, it was just unauthenticated by 2FA)

I am on Fedora Silverblue 38, with uses containerized development environments; and when I run openfortivpn, it doesn't work on the base system. However when I use a container with the correct resolv.conf in a Podman container (using toolbox), the DNS solving works fine.
